crypto/dh/dh_ameth.c

```diff
@@ -540,6 +544,140 @@ static int dh_pkey_param_check(const EVP_PKEY *pkey)
     return DH_check_ex(dh);
 }
 
+static int dh_pkey_is_dirty(const EVP_PKEY *pkey)
+{
+    return pkey->pkey.dh->dirty;
+}
+
+static void dh_pkey_clear_dirty(EVP_PKEY *pkey)
+{
+    pkey->pkey.dh->dirty = 0;
```

Comment:
DH is reference counted which means the same DH object could be shared between multiple EVP_PKEY objects so storing the dirty flag in the dh object itself is problematic, e.g. one EVP_PKEY clears the dirty flag, which means the second one doesn't know that the DH is dirty.

Reply:
Hmmm, and synchronizing the dirty bit to all EVP_PKEYs is problematic as well.  In that case, it's probably better to make it a counter, and let the EVP_PKEY hold a copy that gets updated after an export, if it differed from the legacy key.

This is, BTW, a change for #9312, not this PR

Reply:
> This is, BTW, a change for #9312, not this PR

... or well, partly.

Reply:
That sounds like a neat solution

Reply:
Okie, done, both in #9312 and here